jena-arq/src/main/java/org/apache/jena/sparql/algebra/optimize/TransformPathFlattern.java

```diff
@@ -53,9 +58,53 @@ public TransformPathFlattern(PathCompiler pathCompiler)
     @Override
     public Op transform(OpPath opPath)
     {
+        Op op = transformReduce(opPath, pathCompiler);
+        return op;
+    }
+
+    static Op transformReduce(OpPath opPath, PathCompiler pathCompiler) {
         // Flatten down to triples where possible.
-        PathBlock pattern = pathCompiler.reduce(opPath.getTriplePath()) ;
+        PathBlock pattern = pathCompiler.reduce(opPath.getTriplePath());
         // Any generated paths of exactly one to triple; convert to Op.
-        return PathLib.pathToTriples(pattern) ;
+        Op op = PathLib.pathToTriples(pattern);
+        return transformAlt(op, pathCompiler);
```

Comment:
A request for explanation: 

We seem to have two mechanism for the same thing here - PathCompiler and AltTransformer+transformReduce.

What advantages does this bring?

Reply:
the suggestion of this change is as follows: we leave the original pathCompiler unchanged (`transformReduce`/pathCompiler.reduce) -- it can only flatten a path to PathBlocks and does not handle P_Alt. Then, we add another Transformer (transformAlt/AltTransformer) which only transforms any P_Alt to OpUnion